source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,723 @@
+
+    static assert(functions.length == 2);
+
+    static assert(
+        functions[0].mixture ==
+        q{@system bolts.experimental.refraction.ReturnType!(__traits(getOverloads, Container, "answer")[0]) answer();});
+    static assert(functions[0].index == 0);
+
+    static assert(
+        functions[1].mixture ==
+        q{@system bolts.experimental.refraction.ReturnType!(__traits(getOverloads, Container, "answer")[2]) answer();});
+    static assert(functions[1].index == 2);
+}
+
+private enum isAggregate(T...) =
+    is(T[0] == struct) || is(T[0] == union) || is(T[0] == class)
+    || is(T[0] == interface);
+
+private mixin template replaceAttribute(string Name) {
+    alias Struct = typeof(this);
+    mixin(
+        "Struct copy = {",
+        {
+            string[] mixture;
+            foreach (member; __traits(allMembers, Struct)) {
+                if (__traits(getOverloads, Struct, member).length == 0) {
+                    mixture ~= member ~ ":" ~ (member == Name ? "value" : member);
+                }
+            }
+            return mixture.join(",\n");
+        }(),
+        "};"
+    );
+}
+
+unittest {
+    static struct QA {
+        string question;
+        int answer;
+        QA withQuestion(string value) {
+            mixin replaceAttribute!"question";
+            return copy;
+        }
+    }
+    QA deflt = { answer: 42 };
+    enum question = "How many roads must a man walk down?";
+    auto result = deflt.withQuestion(question);
+    assert(result.question == question);
+    assert(result.answer == 42);
+}
+
+/**
+   A struct capturing all the aspects of a function necessary to produce a
+   string mixin that re-creates the function (excepting the body).
+*/
+
+immutable struct Function {
+
+    /**
+       A string that evaluates to a symbol representing the function in the
+       local context.
+    */
+
+    string localName;
+
+    /**
+       Function name. Initial value: `__traits(identifier, fun)`.
+    */
+
+    string name;
+
+    /**
+       Return a new `Function` object with the `name` attribute set to `value`.
+    */
+
+    Function withName(string value) {
+        mixin replaceAttribute!"name";
+        return copy;
+    }
+
+    ///
+    unittest {
+        pure @nogc int answer();
+        mixin(refract!(answer, "answer").withName("ultimateAnswer").mixture);
+        static assert(
+            __traits(getAttributes, ultimateAnswer) ==
+            __traits(getAttributes, answer));
+    }
+
+    /**
+       Index of function in an arbitrary collection of functions. The index can
+       be set explicitly with `withIndex`. The `refract` overloads that operate
+       on function containers set `index` to the position of the function in
+       the set of overloaded functions in a scope. Default value: -1.
+    */
+
+    int index;
+
+    /**
+       Return a new `Function` object with the `index` attribute set to
+       `value`.
+    */
+
+    Function withIndex(int value) {
```

Comment:
`withOverloadIndex`?

Reply:
> `withOverloadIndex`?

Well...the `index` is initialized to the position of the function in the overload set, but it can be modified, even though the function cannot be moved inside the module ;-) It's a bit the same debate as "name" vs "originalName". 

Also, I found situations where it is more useful to store the function position inside the module, regardless of the overloads. For example, the Mock example could use a `int[]` to keep track of function call counts, and a flat index stored in `index` to increment it.